### ISSUE TITLE
Keybinding to select to left/right bracket

### ISSUE THREAD

**Issue by dsterda** [facelessuser/BracketHighlighter#475]
## Description

Is there support for a keyboard shortcut to select to the beginning / end of a bracket?

For example, given this line of code (the cursor is marked by |):

`print("this is| a test")`

I would like to add a keybinding to select from the cursor position to the end of "test", and another to select to the beginning of "this", e.g.

`print("this is[ a test]")`

and

`print("[this is] a test")`

I found examples to select all of the text between brackets as well as to _move_ the cursor to the beginning/end of brackets, but I wasn't able to determine how to _select_ to the beginning/end of brackets.

## Support Info

- ST ver.: 3143
- Platform: osx
- Arch: x64
- Plugin ver.: 2.27.1
- Install via PC: True
- mdpopups ver.: 3.3.1
- backrefs ver.: 3.5
- markdown ver.: 2.6.9
- pygments ver.: 2.1a0
- jinja2 ver.: 2.8

**Comment by facelessuser** [comment]
There isn't currently such a shortcut.  The way that BracketHighlighter implements its current selections is by using a `bh_module`: https://github.com/facelessuser/BracketHighlighter/blob/master/bh_modules/bracketselect.py.  Which can then be used in a command:

```js
    // Go to left bracket
    {
        "keys": ["ctrl+alt+super+up"],
        "command": "bh_key",
        "args":
        {
            "no_block_mode": null,
            "no_outside_adj": null,
            "lines" : true,
            "plugin":
            {
                "type": ["__all__"],
                "command": "bh_modules.bracketselect",
                "args": {"select": "left"}
            }
        }
    },
    // Go to right bracket
    {
        "keys": ["ctrl+alt+super+down"],
        "command": "bh_key",
        "args":
        {
            "no_block_mode": null,
            "no_outside_adj": null,
            "lines" : true,
            "plugin":
            {
                "type": ["__all__"],
                "command": "bh_modules.bracketselect",
                "args": {"select": "right"}
            }
        }
    },
```

I believe most of this information is covered in the documentation: http://facelessuser.github.io/BracketHighlighter/customize/#run-instance-plugins.

Theoretically, you could write your own specialized command to do exactly what it is you want.

**Comment by dsterda** [comment]
Would it be easy to modify bracketselect.py to support selecting to the beginning/end of the bracket? Or would the best approach be to create a new plugin?

**Comment by facelessuser** [comment]
It is possible to add it to the existing `bh_module`. If we're talking about a pull request, that would make sense.  Ideally in such a way that the code is not hard to follow. If it is going to get super messy, then having a separate module makes more sense.

If it is only for personal use, then doing a separate module probably makes the most sense.

**Comment by dsterda** [comment]
Thanks for the tips, I was able to create my own module. I'd be happy to post it here in case it helps anyone else.

I was wondering if you could help me with one thing; when selecting to the left bracket, the cursor remains at the original position. Is it possible to move the cursor as well as setting the selection?

For example, if the shortcut to select to the left bracket is ctrl+shift+left, and the line is `"thisis|atest"`, pressing ctrl+shift+left followed by shift+right should result in this selection: `"t[hisis]atest"`, but currently results in this selection: `"[thisisa]test"`. The solution (I think) is to position the cursor at `left` before making the selection.

I tried using `self.view.run_command('move_to', {'to': first})`, but that didn't seem to work.

In case it helps, here is the code:
```
    def run(self, edit, name, direction, always_include_brackets=False):
        self.current_left, self.current_right = self.selection[0].begin(), self.selection[0].end()
        first, last = self.left.end, self.right.begin

        if first == self.current_left and last == self.current_right or always_include_brackets:
            first, last = self.select_expand(first, last)

        if direction == 'right':
            first = self.selection[0].begin()
        elif direction == 'left':
            last = self.selection[0].end()

        self.selection = [sublime.Region(first, last)]
        self.refresh_match = True
```

**Comment by facelessuser** [comment]
When selecting, you can control selection direction by reversing the start and end of the selection:

```py
        self.selection = [sublime.Region(last, first)]
```

That should be all you need.

**Comment by dsterda** [comment]
Worked like a charm, thank you!

Here is the full plugin in case it's helpful to anyone else in the future:

```
import BracketHighlighter.bh_plugin as bh_plugin
import sublime


class SelectToBracket(bh_plugin.BracketPluginCommand):

    def run(self, edit, name, direction, always_include_brackets=False):
        self.current_left, self.current_right = self.selection[0].begin(), self.selection[0].end()
        first, last = self.left.end, self.right.begin

        if first == self.current_left and last == self.current_right or always_include_brackets:
            first, last = self.select_expand(first, last)

        if direction == 'right':
            first = self.selection[0].begin()
            self.selection = [sublime.Region(first, last)]
        elif direction == 'left':
            last = self.selection[0].end()
            self.selection = [sublime.Region(last, first)]

        self.refresh_match = True


def plugin():
    return SelectToBracket
```

and keybindings:

```
    {
        "keys": ["ctrl+shift+right"],
        "command": "bh_key",
        "args":
        {
            "no_outside_adj": null,
            "no_block_mode": null,
            "lines" : false,
            "plugin":
            {
                "type": ["__all__"],
                "command": "<path_to_plugin>",
                "args": {"direction": "right"}
            }
        }
    },
    {
        "keys": ["ctrl+shift+left"],
        "command": "bh_key",
        "args":
        {
            "no_outside_adj": null,
            "no_block_mode": null,
            "lines" : false,
            "plugin":
            {
                "type": ["__all__"],
                "command": "<path_to_plugin>",
                "args": {"direction": "left"}
            }
        }
    },
```
